core/src/main/scala/kafka/coordinator/GroupMetadata.scala

```diff
@@ -0,0 +1,190 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.coordinator
+
+import kafka.utils.nonthreadsafe
+
+import java.util.UUID
+
+import org.apache.kafka.common.protocol.Errors
+
+import collection.mutable
+
+private[coordinator] sealed trait GroupState { def state: Byte }
+
+/**
+ * Group is preparing to rebalance
+ *
+ * action: respond to heartbeats with an ILLEGAL GENERATION error code
```

Comment:
And with the FSM, we can write handleXXX in Coordinator in a way that is aligned with the FSM as the following (personally I feel it may make future implementation and reviews easier):

handleXXX:
  1. check coordinator arability and group availability,
  2. if group is available call doXXX.
  3. otherwise proper error code, or create group first then call doXXX.

doXXX:
   switch (group.state)
       case State1: this request should not be received in this state, return some error-code.
       case State2: do some actions based on the group metadata, and probably transit group to another state.
       case StateX: ... etc

And similarly for events like failed consumers, call onYYY:

onYYY:
   switch (group.state)
       case State1: this event should not happen in this state, throw exception.
       case State2: do some actions based on the group metadata, and probably transit group to another state.
       case StateX: ... etc